![Ironhack logo](https://i.imgur.com/1QgrNNw.png)

# Data Preparation

Includes:

- Data ingestion (source: http://casas.wsu.edu/datasets/).

- Data description (Paris, 2+pet, Daily life, 2010)

- Data wrangling and cleaning (Phase 1: general purpose)


In [1]:
#Importing libraries.
import pandas as pd
import numpy as np
import re

In [2]:
#Importing dataset.
df_raw = pd.read_csv('./input/data.csv', sep=' ',index_col=False,header=None,
                      names=['Date','Time','Sensor','Measurement'])

In [3]:
#Visualization.
display(df_raw.head(10),df_raw.tail(10),df_raw.shape,df_raw.dtypes)

Date         Time Sensor Measurement
0  2010-03-31  00:04:46.08   T001        15.5
1  2010-03-31  00:09:48.55   T001          15
2  2010-03-31  00:09:48.60   T003        14.5
3  2010-03-31  00:14:51.03   T001        14.5
4  2010-03-31  00:14:51.08   T003          14
5  2010-03-31  00:16:57.49   M001          ON
6  2010-03-31  00:17:00.54   M001         OFF
7  2010-03-31  00:24:55.93   T003        13.5
8  2010-03-31  00:40:03.35   T001          14
9  2010-03-31  00:55:10.77   T003          13

Date             Time Sensor Measurement
1269255  2010-09-06  23:59:42.217606   P001        0.11
1269256  2010-09-06  23:59:43.223888   P001         0.1
1269257  2010-09-06  23:59:45.234973   P001        0.11
1269258  2010-09-06  23:59:46.241363   P001         0.1
1269259  2010-09-06  23:59:49.256612   P001        0.11
1269260  2010-09-06  23:59:50.263387   P001         0.1
1269261  2010-09-06  23:59:51.269556   P001        0.11
1269262  2010-09-06   23:59:52.27554   P001         0.1
1269263  2010-09-06  23:59:56.305708   P001        0.11
1269264  2010-09-06  23:59:57.301184   P001         0.1

(1269265, 4)

Date           object
Time           object
Sensor         object
Measurement    object
dtype: object

In [4]:
#Hard copy.
df_sensors = df_raw.copy()

In [5]:
#Time string manipulation function (set datetime to 10 minutes window) 1/2.
def time_constrictor1(tiempo_string):
    if len(tiempo_string) >= 9:
        tiempo_string = re.sub(r'\d:\d\d.\d+$','',tiempo_string)
    elif len(tiempo_string) == 8:
        tiempo_string = re.sub(r'\d:\d\d+$','',tiempo_string)    
    elif len(tiempo_string) == 2:
        tiempo_string = tiempo_string+':0'    
    return tiempo_string

In [6]:
#Time string manipulation function (set datetime to 10 minutes window) 2/2.
def time_constrictor2(tiempo_string):
    tiempo_string = tiempo_string+'0'
    return tiempo_string

In [7]:
#Formula application.
df_sensors['Time'] = df_sensors['Time'].apply(time_constrictor1)
df_sensors['Time'] = df_sensors['Time'].apply(time_constrictor2)

In [8]:
#Merge datetime.
df_sensors['DateTime'] = df_sensors['Date'] + ' ' + df_sensors['Time']

In [9]:
#Changing to datetime format and droping string value columns.
df_sensors['DateTime'] = pd.to_datetime(df_sensors['DateTime'])
df_sensors = df_sensors.drop(['Date', 'Time'], axis=1)

In [10]:
#Sensor lists.
s_motion = ['M001','M002','M003','M004','M005','M006','M007','M008','M009','M010','M011','M012','M013',
            'M014','M015','M016','M017','M018','M019','M020','M021','M022','M023','M024','M025','M026',
            'M027','M028','M029']

s_temp = ['T001','T002','T003']

s_doors = ['D001','D002','D003']

s_medicine = ['I001']

s_power = ['P001']

In [11]:
#Separation by sensor for specific treatment.
def data_sensors(df_input,sensor):
    df_stype = df_input[df_input['Sensor'].isin(sensor)]
    return df_stype

### Movement sensors

In [12]:
###Pendiente de refactorizar (aplico fórmula y reseteo index)
df_M001 = data_sensors(df_sensors,['M001']).reset_index(drop=True)
df_M002 = data_sensors(df_sensors,['M002']).reset_index(drop=True)
df_M003 = data_sensors(df_sensors,['M003']).reset_index(drop=True)
df_M004 = data_sensors(df_sensors,['M004']).reset_index(drop=True)
df_M005 = data_sensors(df_sensors,['M005']).reset_index(drop=True)
df_M006 = data_sensors(df_sensors,['M006']).reset_index(drop=True)
df_M007 = data_sensors(df_sensors,['M007']).reset_index(drop=True)
df_M008 = data_sensors(df_sensors,['M008']).reset_index(drop=True)
df_M009 = data_sensors(df_sensors,['M009']).reset_index(drop=True)
df_M010 = data_sensors(df_sensors,['M010']).reset_index(drop=True)
df_M011 = data_sensors(df_sensors,['M011']).reset_index(drop=True)
df_M012 = data_sensors(df_sensors,['M012']).reset_index(drop=True)
df_M013 = data_sensors(df_sensors,['M013']).reset_index(drop=True)
df_M014 = data_sensors(df_sensors,['M014']).reset_index(drop=True)
df_M015 = data_sensors(df_sensors,['M015']).reset_index(drop=True)
df_M016 = data_sensors(df_sensors,['M016']).reset_index(drop=True)
df_M017 = data_sensors(df_sensors,['M017']).reset_index(drop=True)
df_M018 = data_sensors(df_sensors,['M018']).reset_index(drop=True)
df_M019 = data_sensors(df_sensors,['M019']).reset_index(drop=True)
df_M020 = data_sensors(df_sensors,['M020']).reset_index(drop=True)
df_M021 = data_sensors(df_sensors,['M021']).reset_index(drop=True)
df_M022 = data_sensors(df_sensors,['M022']).reset_index(drop=True)
df_M023 = data_sensors(df_sensors,['M023']).reset_index(drop=True)
df_M024 = data_sensors(df_sensors,['M024']).reset_index(drop=True)
df_M025 = data_sensors(df_sensors,['M025']).reset_index(drop=True)
df_M026 = data_sensors(df_sensors,['M026']).reset_index(drop=True)
df_M027 = data_sensors(df_sensors,['M027']).reset_index(drop=True)
df_M028 = data_sensors(df_sensors,['M028']).reset_index(drop=True)
df_M029 = data_sensors(df_sensors,['M029']).reset_index(drop=True)

In [13]:
#Binarization function.
def cat_data(value):
    if value == 'ON':
        value = 1
    elif value == 'OFF':
        value = 0
    return value

In [14]:
###Pendiente de refactorizar (aplico fórmula)
df_M001['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M002['Measurement'] = df_M002['Measurement'].apply(cat_data)
df_M003['Measurement'] = df_M003['Measurement'].apply(cat_data)
df_M004['Measurement'] = df_M004['Measurement'].apply(cat_data)
df_M005['Measurement'] = df_M005['Measurement'].apply(cat_data)
df_M006['Measurement'] = df_M006['Measurement'].apply(cat_data)
df_M007['Measurement'] = df_M007['Measurement'].apply(cat_data)
df_M008['Measurement'] = df_M008['Measurement'].apply(cat_data)
df_M009['Measurement'] = df_M009['Measurement'].apply(cat_data)
df_M010['Measurement'] = df_M010['Measurement'].apply(cat_data)
df_M011['Measurement'] = df_M011['Measurement'].apply(cat_data)
df_M012['Measurement'] = df_M012['Measurement'].apply(cat_data)
df_M013['Measurement'] = df_M013['Measurement'].apply(cat_data)
df_M014['Measurement'] = df_M014['Measurement'].apply(cat_data)
df_M015['Measurement'] = df_M015['Measurement'].apply(cat_data)
df_M016['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M017['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M018['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M019['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M020['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M021['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M022['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M023['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M024['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M025['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M026['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M027['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M028['Measurement'] = df_M001['Measurement'].apply(cat_data)
df_M029['Measurement'] = df_M001['Measurement'].apply(cat_data)

In [15]:
###Pendiente de refactorizar (reduzco dimensión a ventanas de 10 minutos y aplica la media de los valores)
df_M001 = df_M001.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M002 = df_M002.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M003 = df_M003.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M004 = df_M004.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M005 = df_M005.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M006 = df_M006.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M007 = df_M007.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M008 = df_M008.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M009 = df_M009.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M010 = df_M010.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M011 = df_M011.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M012 = df_M012.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M013 = df_M013.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M014 = df_M014.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M015 = df_M015.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M016 = df_M016.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M017 = df_M017.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M018 = df_M018.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M019 = df_M019.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M020 = df_M020.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M021 = df_M021.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M022 = df_M022.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M023 = df_M023.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M024 = df_M024.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M025 = df_M025.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M026 = df_M026.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M027 = df_M027.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M028 = df_M028.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_M029 = df_M029.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})

In [16]:
###Pendiente de refactorizar (etiquetado de columnas)
df_M001.columns=['DateTime','M001']
df_M002.columns=['DateTime','M002']
df_M003.columns=['DateTime','M003']
df_M004.columns=['DateTime','M004']
df_M005.columns=['DateTime','M005']
df_M006.columns=['DateTime','M006']
df_M007.columns=['DateTime','M007']
df_M008.columns=['DateTime','M008']
df_M009.columns=['DateTime','M009']
df_M010.columns=['DateTime','M010']
df_M011.columns=['DateTime','M011']
df_M012.columns=['DateTime','M012']
df_M013.columns=['DateTime','M013']
df_M014.columns=['DateTime','M014']
df_M015.columns=['DateTime','M015']
df_M016.columns=['DateTime','M016']
df_M017.columns=['DateTime','M017']
df_M018.columns=['DateTime','M018']
df_M019.columns=['DateTime','M019']
df_M020.columns=['DateTime','M020']
df_M021.columns=['DateTime','M021']
df_M022.columns=['DateTime','M022']
df_M023.columns=['DateTime','M023']
df_M024.columns=['DateTime','M024']
df_M025.columns=['DateTime','M025']
df_M026.columns=['DateTime','M026']
df_M027.columns=['DateTime','M027']
df_M028.columns=['DateTime','M028']
df_M029.columns=['DateTime','M029']

### Temperature and power sensors

In [17]:
###Pendiente de refactorizar (aplico fórmula y reseteo index)
df_T001 = data_sensors(df_sensors,['T001']).reset_index(drop=True)
df_T002 = data_sensors(df_sensors,['T002']).reset_index(drop=True)
df_T003 = data_sensors(df_sensors,['T003']).reset_index(drop=True)
df_P001 = data_sensors(df_sensors,['P001']).reset_index(drop=True)

In [18]:
#To float function
def num_data(value):
    return float(value)

In [19]:
###Pendiente de refactorizar (aplico fórmula)
df_T001['Measurement'] = df_T001['Measurement'].apply(num_data)
df_T002['Measurement'] = df_T002['Measurement'].apply(num_data)
df_T003['Measurement'] = df_T003['Measurement'].apply(num_data)
df_P001['Measurement'] = df_P001['Measurement'].apply(num_data)

In [20]:
###Pendiente de refactorizar (reduzco dimensión a ventanas de 10 minutos y aplica la media de los valores)
df_T001 = df_T001.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_T002 = df_T002.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_T003 = df_T002.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_P001 = df_P001.groupby('DateTime',as_index=False).agg({"Measurement":"sum"})

In [21]:
###Pendiente de refactorizar (etiquetado de columnas)
df_T001.columns=['DateTime','T001']
df_T002.columns=['DateTime','T002']
df_T003.columns=['DateTime','T003']
df_P001.columns=['DateTime','P001']

### Door sensors

In [22]:
###Pendiente de refactorizar (aplico fórmula y reseteo index)
df_D001 = data_sensors(df_sensors,['D001']).reset_index(drop=True)
df_D002 = data_sensors(df_sensors,['D002']).reset_index(drop=True)
df_D003 = data_sensors(df_sensors,['D003']).reset_index(drop=True)

In [23]:
#Binarization function.
def door_data(value):
    if value == 'OPEN':
        value = 1
    elif value == 'CLOSE':
        value = 0
    return value

In [24]:
###Pendiente de refactorizar (aplico fórmula)
df_D001['Measurement'] = df_D001['Measurement'].apply(door_data)
df_D002['Measurement'] = df_D002['Measurement'].apply(door_data)
df_D003['Measurement'] = df_D003['Measurement'].apply(door_data)

In [25]:
###Pendiente de refactorizar (reduzco dimensión a ventanas de 10 minutos y aplica la media de los valores)
df_D001 = df_D001.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_D002 = df_D002.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})
df_D003 = df_D003.groupby('DateTime',as_index=False).agg({"Measurement":"mean"})

In [26]:
###Pendiente de refactorizar (etiquetado de columnas)
df_D001.columns=['DateTime','D001']
df_D002.columns=['DateTime','D002']
df_D003.columns=['DateTime','D003']

### Data visual analysis

In [27]:
df_T001.head()

DateTime   T001
0 2010-03-31 00:00:00  15.25
1 2010-03-31 00:10:00  14.50
2 2010-03-31 00:40:00  14.00
3 2010-03-31 01:30:00  13.50
4 2010-03-31 01:40:00  15.00

In [28]:
df_P001.tail()

DateTime   P001
8023 2010-09-06 23:10:00   4.62
8024 2010-09-06 23:20:00   6.30
8025 2010-09-06 23:30:00  10.08
8026 2010-09-06 23:40:00  28.56
8027 2010-09-06 23:50:00  28.56

In [29]:
df_P001.shape

(8028, 2)

In [30]:
df_P001.DateTime.min()

Timestamp('2010-05-11 12:10:00')

In [31]:
df_P001.DateTime.max()

Timestamp('2010-09-06 23:50:00')

In [32]:
df_P001.DateTime.max()-df_P001.DateTime.min()

Timedelta('118 days 11:40:00')

In [33]:
df_T001.DateTime.max()-df_T001.DateTime.min()

Timedelta('159 days 23:10:00')

### Dataframe

In [34]:
#Main 'signature matrix' creation.
time_ini = pd.date_range('2010-03-31 00:00:00', periods=23040, freq='10min')
#time_fin = pd.date_range('2010-03-31 00:04:59', periods=230396, freq='1min')
#df_time_ini = pd.DataFrame({'TimeIni':time_ini})
#df_time_fin = pd.DataFrame({'TimeFin':time_fin})
df_full = pd.DataFrame({'DateTime':time_ini})
#df_full = pd.concat([df_time_ini, df_time_fin], axis=1, sort=False)
#df_full = pd.DataFrame({'TimeGap': np.arange(1,230397)},index=time_gap)
df_full['DayWeek'] = df_full['DateTime'].dt.dayofweek
display(df_full.head(),df_full.tail(),df_full.shape)

DateTime  DayWeek
0 2010-03-31 00:00:00        2
1 2010-03-31 00:10:00        2
2 2010-03-31 00:20:00        2
3 2010-03-31 00:30:00        2
4 2010-03-31 00:40:00        2

DateTime  DayWeek
23035 2010-09-06 23:10:00        0
23036 2010-09-06 23:20:00        0
23037 2010-09-06 23:30:00        0
23038 2010-09-06 23:40:00        0
23039 2010-09-06 23:50:00        0

(23040, 2)

In [35]:
#Merge function.
def merge_df(df_main,df_list):
    for df in df_list:
        df_main = pd.merge(df_main,df,how='left',on=['DateTime'])
    return df_main

In [36]:
#Dataframes to merge.
df_all = [df_M001,df_M002,df_M003,df_M004,df_M005,df_M006,df_M007,df_M008,df_M009,df_M010,df_M011,df_M012,
         df_M013,df_M014,df_M015,df_M016,df_M017,df_M018,df_M019,df_M020,df_M021,df_M022,df_M023,df_M024,
         df_M025,df_M026,df_M027,df_M028,df_M029,df_T001,df_T002,df_T003,df_P001,df_D001,df_D002,df_D003]

In [37]:
%%time
#Merge and fill null.
df_01 = merge_df(df_full,df_all).fillna(0)

CPU times: user 582 ms, sys: 4.07 ms, total: 586 ms
Wall time: 585 ms


In [38]:
#Full signature matrix (1/2)
display(df_01.shape,df_01.head())

(23040, 38)

DateTime  DayWeek  M001  M002  M003  M004  M005  M006  M007  \
0 2010-03-31 00:00:00        2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1 2010-03-31 00:10:00        2   0.5   0.0   0.0   0.0   0.0   0.0   0.0   
2 2010-03-31 00:20:00        2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3 2010-03-31 00:30:00        2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4 2010-03-31 00:40:00        2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

   M008  ...   M027  M028  M029   T001  T002  T003  P001  D001  D002  D003  
0   0.0  ...    0.0   0.0   0.0  15.25   0.0   0.0   0.0   0.0   0.0   0.0  
1   0.0  ...    0.0   0.0   0.0  14.50   0.0   0.0   0.0   0.0   0.0   0.0  
2   0.0  ...    0.0   0.0   0.0   0.00   0.0   0.0   0.0   0.0   0.0   0.0  
3   0.0  ...    0.0   0.0   0.0   0.00   0.0   0.0   0.0   0.0   0.0   0.0  
4   0.0  ...    0.0   0.0   0.0  14.00   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 38 columns]

In [39]:
#Full signature matrix (2/2)
display(df_01.dtypes,df_01.DateTime.max()-df_01.DateTime.min())

DateTime    datetime64[ns]
DayWeek              int64
M001               float64
M002               float64
M003               float64
M004               float64
M005               float64
M006               float64
M007               float64
M008               float64
M009               float64
M010               float64
M011               float64
M012               float64
M013               float64
M014               float64
M015               float64
M016               float64
M017               float64
M018               float64
M019               float64
M020               float64
M021               float64
M022               float64
M023               float64
M024               float64
M025               float64
M026               float64
M027               float64
M028               float64
M029               float64
T001               float64
T002               float64
T003               float64
P001               float64
D001               float64
D002               float64
D

Timedelta('159 days 23:50:00')

In [40]:
cols = df_01.columns
counter = 0
for col in cols:
    for x in df_01[col]:
        if x == 0:
            counter += 1
print('Percentage of zeros in f is: {}'.format((counter/(df_01.shape[0]*df_01.shape[1]))*100))

Percentage of zeros in f is: 82.98394097222223


In [41]:
#Data dump
df_01.to_csv(r'./output/data_01.csv')

## Conclusions

Signature matrix characteristics:

- 23040 rows and 38 columns.

- Only numeric values.

- Total time window: 160 days.

- 83% of zero values.
